### DISCUSSION TITLE
return also some kind of error code (enum? sum type?) with error() and err

### DISCUSSION THREAD

**Discussion by vladimir-kraus** [vlang/v#11721] [Ideas]
We can return a string with error("...") and retrieve it with err. But using a string we can only describe the error and we cannot do much else with that. It would be nice to be able to return some kind of enumeration of error codes which we could compare and react differently to different errors. Could not the returned error be a kind of user extensible sum type? For example have a look at "zig" language, it seems to me that they support something like that. Anyway the error handling must remain as simple as possible.

**Comment by JalonSolov** [comment]
V supports returning multiple values already.  No extra types required.

https://github.com/vlang/v/blob/master/doc/docs.md#returning-multiple-values

**Comment by vladimir-kraus** [comment]
Yes I know. But in that case you do not have the luxury of the special error handling syntax with `?` or `or` etc.

**Comment by medvednikov** [comment]
This will be supported. Let me add this to the 0.3 roadmap.

**Comment by Gladear** [comment]
For the time being, your can use `error_with_code` in the function, and `errcode` in the `or` block to differentiate the errors 🙂

**Comment by dumblob** [comment]
There is an effort to do it: https://github.com/vlang/v/issues/9367 (see also the RFC https://github.com/vlang/rfcs/pull/7 which will hopefully be implemented sooner or later).

**Comment by igotfr** [comment]
for reference:
https://github.com/vlang/v/discussions/11834
https://github.com/vlang/v/pull/9043
https://github.com/vlang/v/pull/8924
https://github.com/vlang/v/blob/47884dfd1ff75b5bc898866910710bb75931c7a6/vlib/builtin/option.v
